Add correctness validation for decode_attention test

harvenstar opened the pull request:
Noticed the TODO in test_decode_attention asking for a correctness check. Currently the test just runs decode_attention_fwd without validating output.

  Implemented a float32 reference using PyTorch einsum that:
  - Handles both MHA and GQA configurations (expands KV heads with repeat_interleave when needed)
  - Uses stable softmax (subtract max before exp)
  - Follows the same kv_indices semantics with index_select
  - Validates with allclose (atol=2e-2, rtol=1e-2)

harvenstar committed:
Add correctness test for decode_attention

- Fix typo: correctnesss -> correctness
- Implement float32 reference with stable softmax
- Add proper GQA support with repeat_interleave
- Validate output against PyTorch einsum reference
- Use index_select for correct kv_indices semantics

harvenstar committed:
Implement correctness validation for decode_attention test

Add float32 reference implementation using PyTorch einsum to validate
decode_attention_fwd output. The reference implementation handles both
MHA and GQA configurations correctly.

Changes:
- Use stable softmax with max subtraction
- Handle GQA by expanding KV heads with repeat_interleave
- Use index_select with kv_indices for correct semantics
- Validate with 2e-2 atol and 1e-2 rtol
- Report max absolute error on mismatch

alisonshao commented:
/tag-and-rerun-ci

ispobock commented on test/registered/attention/test_triton_attention_kernels.py:
```diff
@@ -470,10 +470,43 @@ def _test_decode_attention_once(self, B, H_Q, H_KV, D):
+
+        o_ref = torch.empty((B, H_Q, D), dtype=torch.float32, device=device)
+
+        for b in range(B):
```
Can we move it to a torch deocde reference function and call it here?

harvenstar committed:
Refactor: Extract decode_attention_fwd_torch as reusable function

Address review feedback to move inline torch reference implementation into a separate function for better code reusability.

harvenstar committed:
Tighten tolerance threshold for decode_attention test

Reduce atol from 2e-2 to 1e-2 based on empirical error analysis (max observed ~8.3e-3), providing stricter validation while maintaining 20% safety margin.

harvenstar commented on test/registered/attention/test_triton_attention_kernels.py:
```diff
@@ -470,10 +518,18 @@ def _test_decode_attention_once(self, B, H_Q, H_KV, D):
 
+        max_abs_err = (o.to(torch.float32) - o_ref).abs().max().item()
+        self.assertTrue(
+            torch.allclose(o.to(torch.float32), o_ref, atol=2e-2, rtol=1e-2),
```
Tested with atol=1e-3, rtol=1e-3 but it fails. Observed max errors across configs: 3.1e-3, 3.5e-3, 5.8e-3, and 8.3e-3.

  The errors come from the kernel output being bfloat16 (even though reference is fp32). 
  Using atol=1e-2, rtol=1e-2 gives about 20% margin over max observed error while being 2x stricter than the original 2e-2. Does this make sense? Thank you~

harvenstar commented:
<img width="1142" height="642" alt="image" src="https://github.com/user-attachments/assets/bb7cf8f3-99bb-4a74-83ac-3d86fce091cb" />

<img width="1453" height="724" alt="image" src="https://github.com/user-attachments/assets/7559ea6d-57f4-444f-a569-762df0086762" />


@ispobock  The test in this pr, test/registered/attention/test_triton_attention_kernels.py, runs in the PR Test workflow under stage-b-test-small-1-gpu, partition (3). It passed here, link: https://github.com/sgl-project/sglang/actions/runs/20536522211/job/58997026746?pr=15806
